core/embed/trezorhal/optiga/optiga.c

```diff
@@ -109,3 +158,432 @@ bool optiga_random_buffer(uint8_t *dest, size_t size) {
+  if (OPTIGA_SUCCESS != ret) {
+    return false;
+  }
+
+  ret = optiga_parse_metadata(serialized, size, metadata);
+  return OPTIGA_SUCCESS == ret;
+}
+
+static bool write_metadata(uint16_t oid, const optiga_metadata *metadata) {
+  uint8_t serialized[OPTIGA_MAX_METADATA_SIZE] = {0};
+  size_t size = 0;
+
+  optiga_result ret = optiga_serialize_metadata(metadata, serialized,
+                                                sizeof(serialized), &size);
+  if (OPTIGA_SUCCESS != ret) {
+    return false;
+  }
+
+  ret = optiga_set_data_object(oid, true, serialized, size);
+  return OPTIGA_SUCCESS == ret;
+}
+
+bool optiga_set_metadata(uint16_t oid, const optiga_metadata *metadata) {
+  // Read the stored metadata.
+  optiga_metadata metadata_stored = {0};
+  if (!read_metadata(oid, &metadata_stored)) {
+    return false;
+  }
+
+  // If the stored metadata are different, then set them as requested.
+  if (!optiga_compare_metadata(metadata, &metadata_stored)) {
+    if (!write_metadata(oid, metadata)) {
+      return false;
+    }
+
+    // Check that the metadata was written correctly.
+    if (!read_metadata(oid, &metadata_stored)) {
+      return false;
+    }
+    if (!optiga_compare_metadata(metadata, &metadata_stored)) {
+      return false;
+    }
+  }
+
+  // If the metadata aren't locked, then lock them.
+  optiga_metadata metadata_locked = {0};
+  metadata_locked.lcso = OPTIGA_META_LCS_OPERATIONAL;
+  if (!optiga_compare_metadata(&metadata_locked, &metadata_stored)) {
+    if (!write_metadata(oid, &metadata_locked)) {
+      return false;
+    }
+
+    // Check that metadata were locked correctly.
+    if (!read_metadata(oid, &metadata_stored)) {
+      return false;
+    }
+    if (!optiga_compare_metadata(&metadata_locked, &metadata_stored)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+static bool optiga_pin_init_metadata(void) {
+  optiga_metadata metadata = {0};
+
+  // Set metadata for PIN counter reset key / Master secret.
+  memzero(&metadata, sizeof(metadata));
+  metadata.change = OPTIGA_META_ACCESS_ALWAYS;
+  metadata.read = ACCESS_SALTED_PIN;
+  metadata.execute = OPTIGA_META_ACCESS_ALWAYS;
+  metadata.data_type = TYPE_AUTOREF;
+  if (!optiga_set_metadata(OID_PIN_SECRET, &metadata)) {
+    return false;
+  }
+
+  // Set metadata for PIN counter.
+  memzero(&metadata, sizeof(metadata));
+  metadata.change = ACCESS_PIN_SECRET;
+  metadata.read = OPTIGA_META_ACCESS_ALWAYS;
+  metadata.execute = OPTIGA_META_ACCESS_ALWAYS;
+  if (!optiga_set_metadata(OID_PIN_COUNTER, &metadata)) {
+    return false;
+  }
+
+  // Set metadata for salted PIN.
+  memzero(&metadata, sizeof(metadata));
+  metadata.change = ACCESS_PIN_SECRET;
+  metadata.read = OPTIGA_META_ACCESS_NEVER;
+  metadata.execute = ACCESS_PIN_COUNTER;
+  metadata.data_type = TYPE_AUTOREF;
+  if (!optiga_set_metadata(OID_SALTED_PIN, &metadata)) {
+    return false;
+  }
+
+  // TODO consider adding LUC execute access condition for the stretching
+  // secrets.
```

Comment:
I am not sure this is worth doing, considering that Optiga already has the security monitor in place. This could be more trouble than it's worth.

Reply:
See https://github.com/trezor/trezor-firmware/pull/3296#discussion_r1334085131.